### PR TITLE
Add @theme-ui/components

### PR DESCRIPTION
**Pull request by hasparus**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.) 

Failed on my machine with
```
Clean data
Clean logs
Clean output
Using local Definitely Typed at D:\workspace\DefinitelyTyped.
Parsing definitions...
AssertionError [ERR_ASSERTION] [ERR_ASSERTION]: "babel__code-frame" >= "babelify"
    at Object.assertSorted (D:\workspace\DefinitelyTyped\node_modules\types-publisher\src\util\util.ts:640:9)
    at DiskFS.readdir (D:\workspace\DefinitelyTyped\node_modules\types-publisher\src\get-definitely-typed.ts:226:16)
    at Object.parseDefinitions [as default] (D:\workspace\DefinitelyTyped\node_modules\types-publisher\src\parse-definitions.ts:35:98)
    at main (D:\workspace\DefinitelyTyped\node_modules\types-publisher\src\tester\test.ts:24:27)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) 
{
  generatedMessage: false,
  code: 'ERR_ASSERTION',
  actual: false,
  expected: true,
  operator: '=='
}
error Command failed with exit code 1.
```

- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.

---

- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.

```
> npx dts-gen --dt --name @theme-ui/components --template module
npx: installed 59 in 6.878s
Unexpected crash! Please log a bug with the commandline you specified.
ENOENT: no such file or directory, mkdir 'types\@theme-ui\components'

> npx dts-gen --dt --name theme-ui__components --template module -o
npx: installed 59 in 7.202s
Warning: Could not retrieve version/homepage information: HTTP Error 404: Not Found for http://registry.npmjs.org/theme-ui__components
```

I didn't find the docs on how to run dts-gen for namespaced package.
The latter command generated a proper DT directory and I started from that.

---

- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by hasparus** [comment]
I've left components from `Image` to the bottom stubbed as `React.FC`.
Somebody was asking about this two days ago in theme-ui issue, so I figured pushing this to DT ASAP minimizes the chance somebody else ends up doing the same work.

**Comment by hasparus** [comment]
I'm failing `lint` locally with
```
$ dtslint types theme-ui__components
Error: D:/workspace/DefinitelyTyped/types/theme-ui__components/theme-ui__components-tests.tsx:102:10
ERROR: 102:10  expect  TypeScript@next compile error: 
Type '{ label: string; name: string; defaultValue: string; onChange: (e: ChangeEvent<HTMLInputElement>) => void; mx: number[]; }' is not assignable to type 'IntrinsicAttributes & { label: string; name: string; } & Omit<InputProps | PropsWithChildren<InputProps>, "label" | "marginBottom" | "marginLeft" | "marginRight" | "marginTo  Property 'mx' does not exist on type 'IntrinsicAttributes & { label: string; name: string; } & Omit<InputProps | PropsWithChildren<InputProps>, "label" | "marginBottom" | "marginLeft" | "marginRigERROR: 109:10  expect  TypeScript@next compile error: 
Type '{ as: "textarea"; label: string; name: string; rows: number; placeholder: string; onChange: (e: ChangeEvent<HTMLTextAreaElement>) => void; }' is missing the following properties from type 'Omit<DetailedHTMLProps<TextareaHTMLAttributes<HTMLTextAreaElement>, HTMLTextAreaElement>, "label" | "marginBottom" | "marginLeft" | "marginRight" | "marginTop" | "margin" | ... 8 more ... | "as">': required, disabled, hidden, dir, and 264 more.
ERROR: 117:10  expect  TypeScript@next compile error: 
Type '{ as: ForwardRefExoticComponent<Pick<TextareaProps, "required" | "disabled" | "hidden" | "dir" | "form" | "p" | "slot" | "style" | "title" | "color" | "wrap" | "padding" | "key" | "children" | "backgroundColor" | ... 287 more ... | "rows"> & RefAttributes<...>>; ... 5 more ...; px: number[]; }' is missing the following properties from type 'Omit<Pick<TextareaProps, "required" | "disabled" | "hidden" | "dir" | "form" | "p" | "slot" | "style" | "title" | "color" | "wrap" | "padding" | "key" | "children" | "backgroundColor" | ... 287 more ... | "rows"> & RefAttributes<...>, "label" | 
... 13 more ... | "as">': required, disabled, hidden, dir, and 283 more.

    at D:\workspace\DefinitelyTyped\node_modules\dtslint\bin\index.js:193:19
    at Generator.next (<anonymous>)
    at fulfilled (D:\workspace\DefinitelyTyped\node_modules\dtslint\bin\index.js:6:58)
error Command failed with exit code 1.
```

`tsc` in `types/theme-ui__components` doesn't report any error and neither does
`dtslint types theme-ui__components --localTs .\node_modules\typescript\lib` what's pretty weird because this is the same TypeScript version (next 3.8.0-dev.20191214).

**Comment by eknowles** [comment]
looks good 👍

File: types/theme-ui__components/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+ * @see https://theme-ui.com/components/grid
+ */
+export const Grid: ForwardRef<HTMLDivElement, GridProps>;
+
+export interface ButtonProps extends Assign<React.ComponentPropsWithRef<'button'>, BoxOwnProps> {}
+/**
+ * Primitive button component with variants
+ * @see https://theme-ui.com/components/button
+ */
+export const Button: ForwardRef<HTMLButtonElement, BoxProps>;
```
[the remark this answers is not in the document]
Comment (hasparus) [reply types/theme-ui__components/index.d.ts]: omg 🤦‍♂ thanks

**Review by andrewbranch — APPROVED** [review]

**Comment by petrbela** [comment]
@hasparus I think we should also define and export `FlexProps`. Is there a specific reason it was omitted? Happy to make a PR otherwise.

**Comment by hasparus** [comment]
hey @petrbela 👋 I omitted it because it's the same thing as BoxProps (modulo augmentation) but I think it should be exported anyway

We could rename FlexStyleProps to FlexOwnProps and do
```ts
export interface FlexProps extends Assign<React.ComponentProps<'div'>, FlexOwnProps> {}
```
so it can be augmented in user code before adding to DT if theme-ui Flex gets more props

**Pull request merged**